kernel/arch/dreamcast/hardware/spu.c

```diff
@@ -91,6 +93,41 @@ void spu_memload_sq(uintptr_t dst, void *src_void, size_t length) {
     }
 }
 
+void spu_memload_dma(uintptr_t dst, void *src_void, size_t length) {
+
+    uint8_t *src = (uint8_t *)src_void;
+    size_t aligned_len;
+
+    /* Round up to the nearest multiple of 4 */
+    if(length & 3) {
+        length = (length + 4) & ~3;
+    }
+
+    /* Using DMA (or SQ's on fail) for all that is divisible by 32 */
+    aligned_len = length & ~31;
+    length &= 31;
+
+    do {
+        if (spu_dma_transfer(src_void, dst, aligned_len, 1, NULL, NULL) < 0) {
+            if (errno == EINPROGRESS) {
+                thd_pass();
+                continue;
+            }
+            spu_memload_sq(dst, src_void, aligned_len);
```

Comment:
I'm not sure that this kind of fallback should be encoded here. It seems very much like something that would rather be up to the user to take the possible `EINPROGRESS` and decide how to proceed. Especially since the fallback SQ version could *also* be blocked by another SQ transfer already being in process (or cause an unexpected block to pvr DR for instance).

I'll think on this as well, but it may make sense to provide a new version of spu_memload that offers a `type` option (plain G2 write, SQ, or DMA), to have parameters for blocking/nonblocking, etc.

Reply:
Yes, I agree, it doesn't look very good. But returning the EINPROGRESS error here kills the point, since these functions imply replacement with other similar ones, the result of their work is not disputed in the places of their use. What you are talking about here already exists in the form of `spu_dma_transfer()` which is used here. The main point of this function is to guarantee the result, like `spu_memload()` and `spu_memload_sq()`.
I just wanted to replace `spu_memload_sq()` in SFX and stream prefill with this new function. That is, on the contrary, I want to reduce the blocking of SQ's. They will work in extreme cases.

Reply:
Ah sorry, I didn't quite understand you at first.
Yes, I can make a function with a choice of transfer type, but we can't delete the old ones, since backward compatibility is needed, so either add one with types, or one with dma. I chose one with dma, it's the last one anyway, there are simply no other options :)

Reply:
Also I can just remove it, since it's not used yet anywhere.

Reply:
Sorry for the long delay, I had missed that there were updates here when checking back. I see now what you mean that this is meant to guarantee the result and sit above the rest with the existing transfer still in place that could be used in the way I describe.